fix: Windows VSS Event ID 8194

Snshadow opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------

<!--
Describe the changes and their purpose here, as detailed as needed.
-->
Prevent Windows event 8194 log occurring when doing a backup with --use-fs-snapshot flag.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->
Closes #5169 

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

~~- [ ] I have added tests for all code changes.~~
- [ ] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [ ] I'm done! This pull request is ready for review.

Snshadow committed:
fix: Windows VSS Event ID 8194

kapsiR commented on internal/fs/vss_windows.go:
```diff
@@ -810,6 +810,17 @@ func initializeVssCOMInterface() (*ole.IUnknown, error) {
 	}
 
+	// initialize COM security for VSS, this can't be called more then once
+	if err = ole.CoInitializeSecurity(
+		-1,   // Allow *all* VSS writers to communicate back!
+		5,    // RPC_C_AUTHN_LEVEL_PKT_INTEGRITY
+		3,    // RPC_C_IMP_LEVEL_IMPERSONATE
+		0x40, // EOAC_DYNAMIC_CLOAKING
```
According to [the documentation](https://learn.microsoft.com/en-us/windows/win32/vss/security-considerations-for-requestors#setting-requester-default-com-access-check-permission), this should be with the options (from Windows 8 onwards):
 - `RPC_C_AUTHN_LEVEL_PKT_PRIVACY`
 - `RPC_C_IMP_LEVEL_IMPERSONATE`
 - EOAC_STATIC

Snshadow replied:
[EOLE_AUTHENTICATION_CAPABILITIES](https://learn.microsoft.com/en-us/windows/win32/api/objidl/ne-objidl-eole_authentication_capabilities) does not have a declaration for `EOAC_STATIC`, there only exists `EOAC_STATIC_CLOAKING` could be the document is not complete or did they mean `EOAC_STATIC_CLOAKING`?

_enumeration from objidl.h_
```cpp
typedef enum tagEOLE_AUTHENTICATION_CAPABILITIES {
    EOAC_NONE = 0x0,
    EOAC_MUTUAL_AUTH = 0x1,
    EOAC_STATIC_CLOAKING = 0x20,
    EOAC_DYNAMIC_CLOAKING = 0x40,
    EOAC_ANY_AUTHORITY = 0x80,
    EOAC_MAKE_FULLSIC = 0x100,
    EOAC_DEFAULT = 0x800,
    EOAC_SECURE_REFS = 0x2,
    EOAC_ACCESS_CONTROL = 0x4,
    EOAC_APPID = 0x8,
    EOAC_DYNAMIC = 0x10,
    EOAC_REQUIRE_FULLSIC = 0x200,
    EOAC_AUTO_IMPERSONATE = 0x400,
    EOAC_NO_CUSTOM_MARSHAL = 0x2000,
    EOAC_DISABLE_AAA = 0x1000
} EOLE_AUTHENTICATION_CAPABILITIES;
```

Snshadow committed:
fix: use recommended value for calling CoInitializeSecurity

- check if Windows version is previous 6.2(Windows 8)

Snshadow commented on internal/fs/vss_windows.go:
```diff
@@ -811,14 +829,21 @@ func initializeVssCOMInterface() (*ole.IUnknown, error) {
-		// TODO: warn for expected events for VSS failure event
-		fmt.Printf("VSS warning")
+		// TODO warn for expected event logs for VSS IVssWriterCallback failure
+		fmt.Printf("WARNING: failed to initialize security for VSS: %v\n", err)
```
Setting the value to the recommended from the documentation(guessing they meant using `EOAC_STATIC_CLOAKING`), backing up with VSS works properly in Windows 11 and Windows 7(fixing the running issue with [hooking the api](https://github.com/Snshadow/go1.21-over-win7)). Should the value used for CoInitializeSecurity declared in the repository for readability?

Snshadow committed:
chore: add changelog file

kapsiR commented on internal/fs/vss_windows.go:
```diff
@@ -810,6 +828,24 @@ func initializeVssCOMInterface() (*ole.IUnknown, error) {
 	}
 
+	// initialize COM security for VSS, this can't be called more then once
+	var capabilities uint32
+	if isPreWin8 {
+		capabilities = 0x00 // EOAC_NONE
+	} else {
+		capabilities = 0x20 // EOAC_STATIC_CLOAKING
+	}
```
I think we can remove the pre win8 case entirely because [restic doesn't support it since v0.16.2](https://github.com/restic/restic/issues/4636)?

If we want to support it, we have to change more parameters than that. (impersonation level too)

Your thoughts @MichaelEischer ?

kapsiR commented on internal/fs/vss_windows.go:
```diff
@@ -810,6 +828,24 @@ func initializeVssCOMInterface() (*ole.IUnknown, error) {
+		capabilities = 0x20 // EOAC_STATIC_CLOAKING
+	}
+
+	if err = ole.CoInitializeSecurity(
```
Should we add a comment here and a link to the documentation?

e.g.
```go
// Allowing all processes to perform incoming COM calls is not necessarily a security weakness.  
// A requester acting as a COM server, like all other COM servers, always retains the option to authorize its clients on every COM method implemented in its process.
//
// Note that internal COM callbacks implemented by VSS are secured by default.
// Reference: https://learn.microsoft.com/en-us/windows/win32/vss/security-considerations-for-requestors

```

MichaelEischer commented on internal/fs/vss_windows.go:
```diff
@@ -810,6 +828,24 @@ func initializeVssCOMInterface() (*ole.IUnknown, error) {
+	}
+
+	if err = ole.CoInitializeSecurity(
+		-1,   // Allow *all* VSS writers to communicate back!
```
Wrong comment. `CoInitializeSecurity` always sets a NULL security descriptor, which allows communication from all clients. This parameter is `cAuthSvc`. -1 here means that COM gets to decide which authentication services will be registered (https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-coinitializesecurity)

MichaelEischer reviewed (changes requested):
It's still a bit unclear in which situation exactly this change is necessary. Can this occur in default Windows setups? Does it require special user accounts etc.?

Snshadow committed:
update: fix-vss-8194
- remove fallback for previous Windows versions
- add more context for changelog

Snshadow commented:
> It's still a bit unclear in which situation exactly this change is necessary. Can this occur in default Windows setups? Does it require special user accounts etc.?

I have installed clean Windows 11(24H2) and it occurs if `CoInitializeSecurity` is not called in advance. The frequency of these logs varies depending on each execution.
![image](https://github.com/user-attachments/assets/b263fc79-164d-4423-8885-f20fed48d436)

Snshadow commented:
> It's still a bit unclear in which situation exactly this change is necessary. Can this occur in default Windows setups? Does it require special user accounts etc.?

I have installed clean Windows 11(24H2) and it occurs if `CoInitializeSecurity` is not called in advance. The frequency of these logs varies depending on each execution.

![sample](https://github.com/user-attachments/assets/f4138f60-b869-4e25-acdf-305ce9928838)

Snshadow commented:
> It's still a bit unclear in which situation exactly this change is necessary. Can this occur in default Windows setups? Does it require special user accounts etc.?

I have installed clean Windows 11(24H2) and it occurs if `CoInitializeSecurity` is not called in advance. The frequency of these logs varies depending on each execution.

![sample](https://github.com/user-attachments/assets/38fbc497-0614-4345-ba0e-44ddc2dbf794)

Looking into the comment from https://github.com/restic/restic/blob/ed3922ac82f0ea0f3bbb10a8e01ba8ae51e942c6/internal/fs/vss_windows.go#L898-L900, https://microsoft.public.win32.programmer.kernel.narkive.com/aObDj2dD/volume-shadow-copy-backupcomplete-and-vss-e-bad-state#post2
Looks like the occurrence of this event log has been noted before..

MichaelEischer commented on changelog/unreleased/pull-5170:
```diff
@@ -1,6 +1,22 @@
-BugFix: Prevent Windows VSS event log 8194
+Bugfix: Prevent Windows VSS event log 8194 from being created with `restic backup --use-fs-snapshot`
```
The title must be shorter than 80 characters.

Something like `Bugfix: Prevent Windows VSS event log 8194 warnings for backup with VSS snapshot` would work.

MichaelEischer commented on changelog/unreleased/pull-5170:
```diff
@@ -1,6 +1,22 @@
+Bugfix: Prevent Windows VSS event log 8194 from being created with `restic backup --use-fs-snapshot`
 
-Initialize VSS security to prevent VSS 8194 event logs occurring when running `restic backup` with `--use-fs-snapshot` flag in Windows.
+When running `restic backup` with `--use-fs-snapshot` flag in Windows with admin right, event logs like
```
`rights`

MichaelEischer commented on changelog/unreleased/pull-5170:
````diff
@@ -1,6 +1,22 @@
+   Writer Instance ID: {54b151ac-d27d-4628-9cb0-2bc40959f50f}
+```
+are created several times(the backup itself succeeds).
+To resolve this, initialize VSS security by calling `CoInitializeSecurity` following the approach from
````
The changelog is intended for end users, let's just say that the issue has been solved by initializing the necessary security settings from VSS.

MichaelEischer commented on changelog/unreleased/pull-5170:
````diff
@@ -1,6 +1,22 @@
+```
+are created several times(the backup itself succeeds).
+To resolve this, initialize VSS security by calling `CoInitializeSecurity` following the approach from
+https://github.com/microsoft/Windows-classic-samples/blob/main/Samples/VShadowVolumeShadowCopy/cpp/vssclient.cpp#L52-L63 
````
That reference is wrong, the PR uses the options from https://learn.microsoft.com/en-us/windows/win32/vss/security-considerations-for-requestors , and not from that sample. But either way, that's too technical for the changelog. If someone want the details, they can look at the pull request.

MichaelEischer reviewed (changes requested):
> I have installed clean Windows 11(24H2) and it occurs if CoInitializeSecurity is not called in advance. The frequency of these logs varies depending on each execution.

Thanks for checking!

I have a few more nits in the changelog, other than that the PR seem to be ready.

Snshadow committed:
chore: cleanup changelog

MichaelEischer reviewed (approved):
LGTM. Thanks!
